File: fastapi_backend/app/routes/documents.py

from typing import List, Optional

from fastapi import APIRouter, HTTPException, Query, Depends

from app.models.documents import (
    DocumentCreate,
    DocumentRead,
    GetAllDocumentsResponse,
    DocumentUpdate,
    DocumentContentCreate,
    DocumentContentRead,
)
from app.core.services.document_service import DocumentService
from app.core.exceptions import handle_service_exception
from app.api.dependencies import get_document_service

router = APIRouter(tags=["documents"])


@router.post("/", response_model=DocumentRead)
async def create_document(
    document: DocumentCreate, 
    content: Optional[DocumentContentCreate] = None,
    service: DocumentService = Depends(get_document_service)
):
    """
    Create a new document with optional content.
    If content is not provided, only the document metadata is created (useful for folder structure).
    """
    try:
        return await service.create_document(document, content)
    except Exception as e:
        raise handle_service_exception(e)

@router.get("/root", response_model=List[DocumentRead])
async def get_root_documents(
    is_api_ref: Optional[bool] = None,
    service: DocumentService = Depends(get_document_service)
):
    """Get all root-level documents (no parent)"""
    try:
        return await service.get_root_documents(is_api_ref)
    except Exception as e:
        raise handle_service_exception(e)


@router.get("/{doc_id}", response_model=DocumentRead)
async def get_document(
    doc_id: str,
    service: DocumentService = Depends(get_document_service)
):
    """Get document metadata + latest version"""
    try:
        return await service.get_document(doc_id)
    except Exception as e:
        raise handle_service_exception(e)


@router.get("/{doc_id}/versions", response_model=List[DocumentContentRead])
async def list_document_versions(
    doc_id: str,
    service: DocumentService = Depends(get_document_service)
):
    """List all versions of a document"""
    try:
        return await service.list_document_versions(doc_id)
    except Exception as e:
        raise handle_service_exception(e)


@router.get("/{doc_id}/versions/{version_id}", response_model=DocumentContentRead)
async def get_document_version(
    doc_id: str, 
    version_id: str,
    service: DocumentService = Depends(get_document_service)
):
    """Get a specific version (optional: `latest` as alias)"""
    try:
        return await service.get_document_version(doc_id, version_id)
    except Exception as e:
        raise handle_service_exception(e)



@router.get("/{parent_id}/children", response_model=List[DocumentRead])
async def get_child_documents(
    parent_id: str,
    service: DocumentService = Depends(get_document_service)
):
    """Get all child documents"""
    try:
        return await service.get_child_documents(parent_id)
    except Exception as e:
        raise handle_service_exception(e)




@router.get("/{doc_id}/parents", response_model=List[DocumentRead])
async def get_document_parents(
    doc_id: str,
    service: DocumentService = Depends(get_document_service)
):
    """Get all ancestors (full lineage)"""
    try:
        return await service.get_document_parents(doc_id)
    except Exception as e:
        raise handle_service_exception(e)




@router.get("/",response_model=GetAllDocumentsResponse)
async def get_all_documents(
    is_deleted: Optional[bool] = Query(False),
    is_api_ref: Optional[bool] = None,
    parent_id: Optional[str] = None,
    service: DocumentService = Depends(get_document_service)
)-> GetAllDocumentsResponse:
    """Get all documents with complete hierarchy (with optional filters) and its content"""
    try:
        return await service.get_all_documents(is_deleted, is_api_ref, parent_id)
    except Exception as e:
        raise handle_service_exception(e)


@router.put("/{doc_id}", response_model=DocumentRead)
async def update_document(
    doc_id: str, 
    document: DocumentUpdate,
    service: DocumentService = Depends(get_document_service)
):
    """Update document metadata (title, path, etc.) or delete it"""
    try:
        return await service.update_document(doc_id, document)
    except Exception as e:
        raise handle_service_exception(e)


@router.post("/{doc_id}/versions", response_model=DocumentContentRead)
async def create_document_version(
    doc_id: str, 
    content: DocumentContentCreate,
    service: DocumentService = Depends(get_document_service)
)-> DocumentContentRead:
    """Create a new version for a document (and update latest version)"""
    try:
        return await service.create_document_version(doc_id, content)
    except Exception as e:
        raise handle_service_exception(e)

File: nextjs-frontend/app/api/documents.ts
import { apiClient } from "./client";

export interface Document {
  id: string;
  name: string;
  title: string;
  path: string;
  is_api_ref: boolean;
  parent_id?: string;
  current_version_id?: string;
  is_deleted: boolean;
  created_at: string;
  updated_at: string;
  children?: Document[];
}

export interface DocumentContent {
  version: string;
  document_id: string;
  language: string;
  markdown_content: string;
  keywords_array: string[];
  urls_array: string[];
  summary: string;
  embedding?: number[];
  created_at: string;
  updated_at: string;
}

export interface CreateDocumentRequest {
  name: string;
  title: string;
  path: string;
  is_api_ref?: boolean;
  parent_id?: string;
  language?: string;
  markdown_content?: string;
}

export interface UpdateDocumentRequest {
  name?: string;
  title?: string;
  path?: string;
  is_api_ref?: boolean;
  parent_id?: string;
  is_deleted?: boolean;
}

export interface CreateVersionRequest {
  language: string;
  markdown_content: string;
}

/**
 * API service for document operations
 */
export const documentsApi = {
  /**
   * Get all documents with hierarchy and content
   */
  async getAllDocuments(): Promise<any> {
    return apiClient.get("/api/documents/");
  },

  /**
   * Get root-level documents
   */
  async getRootDocuments(): Promise<Document[]> {
    return apiClient.get("/api/documents/root");
  },

  /**
   * Get document by ID with metadata and latest version
   */
  async getDocument(documentId: string): Promise<Document> {
    return apiClient.get(`/api/documents/${documentId}`);
  },

  /**
   * Get child documents of a parent document
   */
  async getChildDocuments(parentId: string): Promise<Document[]> {
    return apiClient.get(`/api/documents/${parentId}/children`);
  },

  /**
   * Get document lineage (parents chain)
   */
  async getDocumentParents(documentId: string): Promise<Document[]> {
    return apiClient.get(`/api/documents/${documentId}/parents`);
  },

  /**
   * Create a new document
   */
  async createDocument(data: CreateDocumentRequest): Promise<Document> {
    return apiClient.post("/api/documents/", data);
  },

  /**
   * Update document metadata
   */
  async updateDocument(documentId: string, data: UpdateDocumentRequest): Promise<Document> {
    return apiClient.put(`/api/documents/${documentId}`, data);
  },

  /**
   * Create a new version of a document
   */
  async createDocumentVersion(documentId: string, data: CreateVersionRequest): Promise<DocumentContent> {
    return apiClient.post(`/api/documents/${documentId}/versions`, data);
  },
};

/**
 * Cache management for documents
 */
interface CacheEntry {
  data: any;
  timestamp: number;
}

class DocumentsCache {
  private cache: Map<string, CacheEntry> = new Map();
  private readonly TTL = 5 * 60 * 1000; // 5 minutes cache duration

  set(key: string, data: any): void {
    this.cache.set(key, {
      data,
      timestamp: Date.now(),
    });
  }

  get(key: string): any | null {
    const entry = this.cache.get(key);

    if (!entry) {
      return null;
    }

    // Check if cache is expired
    if (Date.now() - entry.timestamp > this.TTL) {
      this.cache.delete(key);
      return null;
    }

    return entry.data;
  }

  clear(): void {
    this.cache.clear();
  }

  invalidate(key: string): void {
    this.cache.delete(key);
  }
}

const documentsCache = new DocumentsCache();

/**
 * Fetch documents with caching
 */
export async function fetchDocumentsWithCache(): Promise<any> {
  const cacheKey = "documents_all";

  // Check cache first
  const cached = documentsCache.get(cacheKey);
  if (cached) {
    console.log("Returning cached documents");
    return cached;
  }

  // Fetch from API if not in cache
  console.log("Fetching documents from API");
  const data = await documentsApi.getAllDocuments();

  // Store in cache
  documentsCache.set(cacheKey, data);

  return data;
}

/**
 * Invalidate document cache after updates
 */
export function invalidateDocumentCache(): void {
  const cacheKey = "documents_all";
  documentsCache.invalidate(cacheKey);
  console.log("Document cache invalidated");
}

/**
 * Force refresh documents cache
 */
export async function refreshDocumentCache(): Promise<any> {
  // First invalidate the cache
  invalidateDocumentCache();

  // Then fetch fresh data
  return await fetchDocumentsWithCache();
}